huffman/src/bit_span.hpp

```diff
@@ -113,5 +115,57 @@ class bit_span : public std::ranges::view_interface<bit_span>
   {
     return iterator{*this, bit_offset_ + bit_size_};
   };
+
+  template <class T>
+  constexpr auto pop() -> T
+  {
+    // TODO
+    // assert byte aligned
+    T res;
+    std::memcpy(&res, data_, sizeof(T));
+    std::advance(data_, sizeof(T));
+    if (std::endian::native == std::endian::big) {
+      res = std::byteswap(res);
+    }
+    return res;
+  }
+
+  constexpr auto pop_8() -> std::uint8_t { return pop<std::uint8_t>(); }
+
+  constexpr auto pop_16() -> std::uint16_t { return pop<std::uint16_t>(); }
+
+  /// Consumes the given number of bits. Basically advances the start of the
+  /// view.
+  ///
+  /// @pre n <= std::ranges::size(this)
+  constexpr auto consume(std::size_t n) -> void
+  {
+    assert(n <= bit_size_);
+    bit_size_ -= n;
+    bit_offset_ += n % CHAR_BIT;
+    if (bit_offset_ == CHAR_BIT) {
+      bit_offset_ = 0;
+      n += CHAR_BIT;
+    }
+    std::advance(data_, n / CHAR_BIT);
+  }
+
+  /// Consumes bits until the start is aligned to a byte boundary.
+  constexpr auto consume_to_byte_boundary() -> void
```

Comment:
maybe return the number of bits consumed?

Reply:
The caller doesn't need it at the moment, so I'd rather keep the interface simpler.